2021/day12/Day12.java

```diff
@@ -0,0 +1,109 @@
+    static HashMap<String, Integer> visitedCavesPartTwo = new HashMap<>();
+
+    static boolean isSmallCave(String cave) {
+        return cave.equals(cave.toLowerCase());
+    }
+
+    static void readInput() {
+        Scanner sc;
+        try {
+            sc = new Scanner(new File("data.txt"));
+            while (sc.hasNextLine()) {
+                String[] caves = sc.nextLine().split("-");
+                String firstCave = caves[0];
+                String secondCave = caves[1];
+
+                if (neighbourCaves.containsKey(firstCave)) {
+                    neighbourCaves.get(firstCave).add(secondCave);
+                } else {
+                    neighbourCaves.put(firstCave, new ArrayList<>(List.of(secondCave)));
+                }
+
+                if (neighbourCaves.containsKey(secondCave)) {
+                    neighbourCaves.get(secondCave).add(firstCave);
+                } else {
+                    neighbourCaves.put(secondCave, new ArrayList<>(List.of(firstCave)));
+                }
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went horribly wrong: " + e);
+        }
+    }
+
+    static void findPath(String currentCave, String currentPath, HashMap<String, Integer> visitedCaves) {
+        currentPath += "," + currentCave;
+
+        if (currentCave.equals("end")) {
+            paths.add(currentPath);
+            return;
+        }
+
+        if (isSmallCave(currentCave)) {
+            visitedCaves.put(currentCave, 1);
+        }
+
+        for (String neighbourCave : neighbourCaves.get(currentCave)) {
+            if (!neighbourCave.equals("start") && !visitedCaves.containsKey(neighbourCave)) {
+                findPath(neighbourCave, currentPath, new HashMap<>(visitedCaves));
+            }
+        }
+    }
+
+    static void findPathPartTwo(String currentCave, String currentPath, HashMap<String, Integer> visitedCavesPartTwo, boolean wasAnyCaveVisitedTwice) {
+        currentPath += "," + currentCave;
+
+        if (currentCave.equals("end")) {
+            pathsPartTwo.add(currentPath);
+            return;
+        }
+
+        if (isSmallCave(currentCave)) {
+            if (visitedCavesPartTwo.containsKey(currentCave)) {
+                visitedCavesPartTwo.put(currentCave, visitedCavesPartTwo.get(currentCave) + 1);
+            } else {
+                visitedCavesPartTwo.put(currentCave, 1);
+            }
+
+            if (visitedCavesPartTwo.get(currentCave) >= 2 && wasAnyCaveVisitedTwice) {
+                return;
+            }
+
+            if (visitedCavesPartTwo.get(currentCave) == 2 && !wasAnyCaveVisitedTwice) {
+                wasAnyCaveVisitedTwice = true;
+            }
+        }
+
+        for (String neighbourNode : neighbourCaves.get(currentCave)) {
+            if (!neighbourNode.equals("start") && visitedCavesPartTwo.getOrDefault(neighbourNode, 0) == 0
```

Comment:
I don't think you need the count check here, right? You already do the check in lines 69-83 and return if it's violated.

Also, `!neighbourNode.equals("start")` appears twice.